File: 코드/ch5/clip6/boj14888/main.py
import sys

PLUS = 0
MINUS = 1
MUL = 2
DIV = 3

def solve(index, total):
    global max_val, min_val
    if index == n:
        max_val = max(max_val, total)
        min_val = min(min_val, total)
        return
    
    for i in range(4):
        if operators[i] > 0:
            operators[i] -= 1
            if i == PLUS:
                solve(index + 1, total + numbers[index])
            elif i == MINUS:
                solve(index + 1, total - numbers[index])
            elif i == MUL:
                solve(index + 1, total * numbers[index])
            elif i == DIV:
                solve(index + 1, int(total / numbers[index]))  # 정수 나눗셈
            operators[i] += 1

n = int(sys.stdin.readline())
numbers = list(map(int, sys.stdin.readline().split()))
operators = list(map(int, sys.stdin.readline().split()))

max_val = -sys.maxsize
min_val = sys.maxsize

solve(1, numbers[0])

print(max_val)
print(min_val)

File: 코드/ch5/clip6/boj14888/Main.java
import java.util.Scanner;

class Main {
    public static final int PLUS = 0;
    public static final int MINUS = 1;
    public static final int MUL = 2;
    public static final int DIV = 3;

    public static int n;
    public static int[] numbers;
    public static int[] operators = new int[4];

    public static int max = Integer.MIN_VALUE;
    public static int min = Integer.MAX_VALUE;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        numbers = new int[n];
        for (int i = 0; i < n; i++) {
            numbers[i] = sc.nextInt();
            ;
        }
        for (int i = 0; i < 4; i++) {
            operators[i] = sc.nextInt();
        }
        solve(1, numbers[0]);
        System.out.println(max);
        System.out.println(min);
    }

    static void solve(int index, int sum) {
        // base case
        if (index == n) {
            if (sum > max) max = sum;
            if (sum < min) min = sum;
            return;
        }
        // recursive case
        for (int i = 0; i < 4; i++) {
            if (operators[i] > 0) {
                operators[i]--;
                switch (i) {
                    case PLUS -> solve(index + 1, sum + numbers[index]);
                    case MINUS -> solve(index + 1, sum - numbers[index]);
                    case MUL -> solve(index + 1, sum * numbers[index]);
                    case DIV -> solve(index + 1, sum / numbers[index]);
                }
                operators[i]++;
            }
        }
    }
}